### PR TITLE
Add infrastructure for offline pretraining of Bandits

### LINKED ISSUE #81: Implement warm-start of adaptive mutation agent
Не в этом PR-е, но в будущем: было бы полезно реализовать подгрузку пред-обученного агента (и, соответственно, его сохранение). Чтобы он стартовал с какого-то warm start-а.

_Originally posted by @nicl-nno in https://github.com/aimclub/GOLEM/issues/78#issuecomment-1500718305_

### PR DESCRIPTION
**Pull request by gkirgizov**
Will enable experiments with offline pretraining of models (both bandit and surrogate also) to compare with baseline GOLEM performance.

Changes:
- Provides infrastructure for collecting histories (`HistoryReader`) and training models separately from optimization (`AgentTrainer`). Those two can work in tnadem with `ExperimentLauncher` and `ExperiementAnalyzer`.
- Extends `ExperienceBuffer` to handle individual trajectories (instead of just steps).
- Extends experiment code in `molecule_search/experiment.py` to run with offline pretraining agent on histories.

Minor changes:
- Additionally, minor fix in `MolAdapter` & `MolGraph`
- Add `__bool__` to Fitness for simpler checks

How to use offline training:
1. Collect histories to some directory using `ExperimentLauncher`
2. Create optimizer & Pretrain mutation agent on these histories using `HistoryReader` and `AgentTrainer`
3. Optionally, validate the Agent on validation set of histories
4. Run optimization with pretrained agent
5. Analyze results of experiments with `ExperiementAnalyzer`

Example in `molecule_search/experiment.py`

### CODE REVIEW

**Comment by gkirgizov** [comment]
@YamLyubov , please check commit `[Fix error with content loading of molecules](https://github.com/aimclub/GOLEM/pull/138/commits/d9399bf961cc7a06c849d8e9b9a65e8055c7c1fe)` in particular

File: golem/core/optimisers/adaptive/agent_trainer.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                 mutation_operator: Mutation,
+                 agent: Optional[OperatorAgent] = None,
+                 ):
+        self._log = default_log(self)
+        self.agent = agent if agent is not None else mutation_operator.agent
+        self.mutation = mutation_operator
+        self.objective = objective
+        self._adapter = self.mutation.graph_generation_params.adapter
+
+    def fit(self, histories: Iterable[OptHistory], validate_each: int = -1) -> OperatorAgent:
```
Comment (maypink) [inline golem/core/optimisers/adaptive/agent_trainer.py:45]: было бы славно заиметь и здесь докстрингу на тему того, что validate_each -- разбиение каждой X истории для  валидации

File: golem/core/optimisers/adaptive/history_collector.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+            # Run optimizer setup
+            optimizer = self.optimizer_cls(objective=self.objective,
+                                           initial_graphs=self.initial_graphs,
+                                           requirements=graph_requirements,
+                                           graph_generation_params=self.graph_generation_params,
+                                           graph_optimizer_params=self.graph_optimizer_params)
+
+            history_filename = f'history_trial{i}.json'
+            optimizer.history.save(self.save_path / history_filename)
```
[the remark this answers is not in the document]
Comment (maypink) [reply golem/core/optimisers/adaptive/history_collector.py:79]: и хочется пример использования во всяком случае. а то непонятно, почему `HistoryCollector` runs Optimizer to collect histories, я бы ожидала, что класс будет собирать опыт из указанной директории, как это делает `HistoryReader`

**Review by maypink — CHANGES_REQUESTED** [review]

File: golem/core/optimisers/adaptive/experience_buffer.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        self._prev_pop = set()
+        self._next_pop = set()
+
+    @staticmethod
+    def unroll_action_step(result: Individual) -> TrajectoryStep:
+        """Unrolls individual's history to get its source individual, action and resulting reward."""
+        if not result.parent_operator or result.parent_operator.type_ != 'mutation':
+            return None, None, np.nan
+        source_ind = result.parent_operator.parent_individuals[0]
+        action = result.parent_operator.operators[0]
```
Comment (maypink) [inline golem/core/optimisers/adaptive/experience_buffer.py:38]: бтв когда я писала подобную штуку для MetaFEDOT, столкнулась с тем, что мутации нужно хранить в ExperienceBuffer как объекты, а не как строки. При подгрузке из историй вроде бы получаются именно строки, с этим не возникает проблем?

**Review by maypink — CHANGES_REQUESTED** [review]

File: examples/molecule_search/experiment.py [hunk 4]
```diff
@@ -208,10 +224,11 @@ def plot_experiment_comparison(experiment_ids: Sequence[str], metric_id: int = 0
 
 if __name__ == '__main__':
     run_experiment(molecule_search_setup,
-                   adaptive_kind=MutationAgentTypeEnum.random,
```
[the remark this answers is not in the document]
Comment (maypink) [reply examples/molecule_search/experiment.py:211]: поменяла LearningPolicy у бандита, теперь все ок

**Review by YamLyubov — APPROVED** [review]

**Review by maypink — APPROVED** [review]
